File: src/main/java/com/rs/java/game/player/actions/skills/crafting/LeatherCrafting.java
package com.rs.java.game.player.actions.skills.crafting;

import com.rs.core.cache.defintions.ItemDefinitions;
import com.rs.java.game.item.Item;
import com.rs.java.game.player.Player;
import com.rs.java.game.player.Skills;
import com.rs.java.game.player.actions.Action;
import com.rs.java.game.player.actions.skills.crafting.leather.LeatherData;
import com.rs.java.game.player.actions.skills.crafting.leather.LeatherProduct;
import com.rs.java.game.player.actions.skills.crafting.leather.ReqItem;

public class LeatherCrafting extends Action {

	public static final int THREAD = 1734;
	public static final int DUNG_NEEDLE = 17446;
	public static final int NORMAL_NEEDLE = 1733;
	private static final int CRAFT_ANIMATION = 1249;

	private int crafted = 0;

	private LeatherData data;
	private LeatherProduct product;
	private int quantity;

	public LeatherCrafting(LeatherData data, int option, int quantity) {
		this.data = data;
		this.product = data.getProducts()[option];
		this.quantity = quantity;
	}

	@Override
	public boolean start(Player player) {
		return check(player);
	}

	private boolean check(Player player) {

		if (player.getSkills().getLevel(Skills.CRAFTING) < product.getLevel()) {
			player.message("You need a Crafting level of " + product.getLevel() + ".");
			return false;
		}
		if (!player.getToolbelt().contains(NORMAL_NEEDLE) && !player.getInventory().containsOneItem(NORMAL_NEEDLE)) {
			player.message("You don't have a needle to craft with.");
			return false;
		}
		if (!player.getInventory().containsItem("item.thread", 1)) {
			player.message("You don't have any thread to craft with.");
			return false;
		}

		StringBuilder missing = new StringBuilder();

		for (ReqItem req : product.getRequirements()) {
			int have = player.getInventory().getAmountOf(req.getId());

			if (have < req.getAmount()) {
				if (!missing.isEmpty())
					missing.append(", ");

				missing.append(req.getAmount())
						.append(" ")
						.append(ItemDefinitions.getItemDefinitions(req.getId()).getName());
			}
		}

		if (!missing.isEmpty()) {
			player.message("You need: " + missing + ".");
			return false;
		}

		return true;
	}

	public static LeatherData getLeatherData(int itemId) {
		for (LeatherData data : LeatherData.values()) {
			if (data.getBaseLeather() == itemId)
				return data;
		}
		return null;
	}

	public static LeatherData getLeatherData(Item used, Item usedWith) {
		int id1 = used.getId();
		int id2 = usedWith.getId();

		boolean hasNeedle =
				id1 == DUNG_NEEDLE || id2 == DUNG_NEEDLE ||
						id1 == NORMAL_NEEDLE || id2 == NORMAL_NEEDLE;

		if (!hasNeedle)
			return null;

		for (LeatherData data : LeatherData.values()) {
			int leatherId = data.getBaseLeather();
			if (leatherId == id1 || leatherId == id2)
				return data;
		}
		return null;
	}

	@Override
	public boolean process(Player player) {

		if (quantity <= 0)
			return false;
		player.animate(CRAFT_ANIMATION);
		if (player.getSkills().getLevel(Skills.CRAFTING) < product.getLevel()) {
			player.message("You need a Crafting level of " + product.getLevel() + " to continue crafting this.");
			return false;
		}
		if (!player.getInventory().containsItem("item.thread", 1)) {
			player.message("You ran out of thread.");
			return false;
		}

		if (product.getRequirements().length == 0) {
			if (!player.getInventory().containsItem(data.getBaseLeather(), 1)) {
				String name = ItemDefinitions.getItemDefinitions(data.getBaseLeather()).getName();
				player.message("You have run out of " + name + ".");
				return false;
			}
			return true;
		}

		for (ReqItem req : product.getRequirements()) {
			if (!player.getInventory().containsItem(req.getId(), req.getAmount())) {
				String name = ItemDefinitions.getItemDefinitions(req.getId()).getName();
				player.message("You have run out of " + name + ".");
				return false;
			}
		}

		return true;
	}


	@Override
	public int processWithDelay(Player player) {

		Integer crafted = (Integer) player.getTemporaryAttributtes().get("THREAD_CRAFT_PROGRESS");
		if (crafted == null)
			crafted = 0;

		quantity--;

		if (product.getRequirements().length == 0) {
			player.getInventory().deleteItem(data.getBaseLeather(), 1);
		} else {
			for (ReqItem req : product.getRequirements()) {
				player.getInventory().deleteItem(req.getId(), req.getAmount());
			}
		}

		crafted++;

		if (crafted % 5 == 0) {
			player.getInventory().deleteItem(THREAD, 1);
		}

		player.getTemporaryAttributtes().put("THREAD_CRAFT_PROGRESS", crafted);

		player.getInventory().addItem(product.getId(), 1);
		player.getSkills().addXp(Skills.CRAFTING, product.getXp());

		return 3;
	}



	@Override
	public void stop(Player player) {
		setActionDelay(player, 3);
	}
}

File: src/raynna/game/player/actions/skills/crafting/leather/LeatherCrafting.kt
package raynna.game.player.actions.skills.crafting.leather

import raynna.core.cache.defintions.ItemDefinitions
import raynna.game.item.Item
import raynna.game.player.Player
import raynna.game.player.Skills
import raynna.game.player.actions.Action

class LeatherCrafting(
    private val data: LeatherData,
    option: Int,
    private var quantity: Int,
) : Action() {
    private val product: LeatherProduct = data.products[option]

    override fun start(player: Player): Boolean = check(player)

    private fun check(player: Player): Boolean {
        if (player.skills.getLevel(Skills.CRAFTING) < product.level) {
            player.message("You need a Crafting level of ${product.level}.")
            return false
        }

        if (!player.hasTool(NORMAL_NEEDLE)) {
            player.message("You don't have a needle to craft with.")
            return false
        }

        if (!player.inventory.containsItem("item.thread", 1)) {
            player.message("You don't have any thread to craft with.")
            return false
        }

        val missing = StringBuilder()

        for (req in product.requirements) {
            val have = player.inventory.getAmountOf(req.getId())

            if (have < req.amount) {
                if (missing.isNotEmpty()) {
                    missing.append(", ")
                }

                missing
                    .append(req.amount)
                    .append(" ")
                    .append(ItemDefinitions.getItemDefinitions(req.getId()).name)
            }
        }

        if (missing.isNotEmpty()) {
            player.message("You need: $missing.")
            return false
        }

        return true
    }

    override fun process(player: Player): Boolean {
        if (quantity <= 0) {
            return false
        }

        player.animate(CRAFT_ANIMATION)

        if (player.skills.getLevel(Skills.CRAFTING) < product.level) {
            player.message("You need a Crafting level of ${product.level} to continue crafting this.")
            return false
        }

        if (!player.inventory.containsItem("item.thread", 1)) {
            player.message("You ran out of thread.")
            return false
        }

        if (product.requirements.isEmpty()) {
            val base = data.getBaseLeather()

            if (!player.inventory.containsItem(base, 1)) {
                val name = ItemDefinitions.getItemDefinitions(base).name
                player.message("You have run out of $name.")
                return false
            }

            return true
        }

        for (req in product.requirements) {
            if (!player.inventory.containsItem(req.getId(), req.amount)) {
                val name = ItemDefinitions.getItemDefinitions(req.getId()).name
                player.message("You have run out of $name.")
                return false
            }
        }

        return true
    }

    override fun processWithDelay(player: Player): Int {
        var crafted = player.temporaryAttributtes["THREAD_CRAFT_PROGRESS"] as? Int ?: 0

        quantity--

        if (product.requirements.isEmpty()) {
            player.inventory.deleteItem(data.getBaseLeather(), 1)
        } else {
            for (req in product.requirements) {
                player.inventory.deleteItem(req.getId(), req.amount)
            }
        }

        crafted++

        if (crafted % 5 == 0) {
            player.inventory.deleteItem(THREAD, 1)
        }

        player.temporaryAttributtes["THREAD_CRAFT_PROGRESS"] = crafted

        player.inventory.addItem(product.getId(), 1)
        player.skills.addXp(Skills.CRAFTING, product.xp)

        return 3
    }

    override fun stop(player: Player) {
        setActionDelay(player, 3)
    }

    companion object {
        const val THREAD = 1734
        const val DUNG_NEEDLE = 17446
        const val NORMAL_NEEDLE = 1733

        private const val CRAFT_ANIMATION = 1249

        fun getLeatherData(itemId: Int): LeatherData? = LeatherData.entries.firstOrNull { it.getBaseLeather() == itemId }

        fun getLeatherData(
            used: Item,
            usedWith: Item,
        ): LeatherData? {
            val id1 = used.id
            val id2 = usedWith.id

            if (id1 == id2) {
                return null
            }

            val hasNeedle =
                id1 == DUNG_NEEDLE || id2 == DUNG_NEEDLE ||
                    id1 == NORMAL_NEEDLE || id2 == NORMAL_NEEDLE

            for (data in LeatherData.entries) {
                val base = data.getBaseLeather()

                if (hasNeedle && (base == id1 || base == id2)) {
                    return data
                }

                for (p in data.products) {
                    var found1 = false
                    var found2 = false

                    for (r in p.requirements) {
                        val rid = r.getId()
                        if (rid == id1) found1 = true
                        if (rid == id2) found2 = true
                    }

                    if (found1 && found2) {
                        return data
                    }
                }
            }

            return null
        }
    }
}
